### PR TITLE
[Tabs/Framework] Cache tabs icon 

### PR DESCRIPTION
**Pull request by AnshulMalik**
Fixes Issue: #6725 

### Summary of Changes

* Change tabs structure from [url] to [{url, framework}]
* Cache the framework 

### Test Plan
- [ ] need to do a little more testing  
- [ ] Initial load takes time, make it performant
- [ ] Upon selecting a tab, the tab becomes first tabIndex

Test it:
- Go to [ahfarmer.github.io](https://ahfarmer.github.io/calculator/)
- Open a few components

tabs

### CODE REVIEW

File: src/actions/tabs.js [hunk 1]
```diff
@@ -16,17 +16,30 @@ import {
...
+    framework
+  };
+}
+
+export function addTab(url: string, framework: string): Action {
   return {
     type: "ADD_TAB",
     url,
-    tabIndex
+    framework
```
[the remark this answers is not in the document]
Comment (AnshulMalik) [reply src/actions/tabs.js]: Umm, I think we were not using tabIndex with updateTab anywhere. Keeping it makes more sense, but not being used is another side. Lemme know if we should keep it.

**Comment by darkwing** [comment]
Hello @AnshulMalik !  I made a bunch of updates to fix half of the flow issues and one of the test fixes.  We should sync up to figure out the rest, or leave some comments on here if you have questions or comments.  Great work so far!

**Comment by AnshulMalik** [comment]
That's great! @darkwing Thanks for the help. I'll have a look soon.

File: src/actions/sources/select.js [hunk 2]
```diff
@@ -128,7 +130,10 @@ export function selectLocation(
       source = getSourceFromId(getState(), location.sourceId);
     }
 
-    dispatch(addTab(source.url, 0));
+    await dispatch(addTab(source.url));
```
Comment (darkwing) [inline src/actions/sources/select.js]: - This is the line that's breaking mochitests.
- This also appears to be the line that's causing the perceived performance slowdown.

File: src/actions/sources/select.js [hunk 3]
```diff
@@ -128,7 +130,12 @@ export function selectLocation(
       source = getSourceFromId(getState(), location.sourceId);
     }
 
-    dispatch(addTab(source.url, 0));
+    dispatch(addTab(source.url));
+    const framework = getFramework(getTabs(getState()), source.url);
+    if (framework) {
+      dispatch(updateTab(source.url, framework));
```
Comment (darkwing) [inline src/actions/sources/select.js]: @AnshulMalik Could you describe the need for this update call?

**Comment by darkwing** [comment]
@jasonLaster This is far enough to get a good review on

File: src/actions/sources/select.js (hunk 3 above)
Comment (AnshulMalik) [reply src/actions/sources/select.js]: Just realised, that's not needed anymore :)

File: src/actions/types/SourceAction.js [hunk 4]
```diff
@@ -49,10 +49,6 @@ export type SourceAction =
         +isBlackBoxed: boolean
       |}
     >
-  | {|
-      +type: "ADD_TAB",
```
Comment (AnshulMalik) [inline src/actions/types/SourceAction.js]: Not sure if they are needed anywhere

File: src/reducers/tabs.js [hunk 5]
```diff
@@ -58,18 +59,25 @@ export function removeSourcesFromTabList(tabs: TabList, urls: TabList) {
...
  */
-function updateTabList(tabs: TabList, url: string, newIndex: ?number) {
-  const currentIndex = tabs.indexOf(url);
+function updateTabList(tabs: TabList, { url, framework = "" }) {
+  const currentIndex = tabs.findIndex(tab => tab.url == url);
   if (currentIndex === -1) {
-    tabs = [url, ...tabs];
-  } else if (newIndex !== undefined) {
-    tabs = move(tabs, currentIndex, newIndex);
+    tabs = [{ url, framework }, ...tabs];
```
[the remark this answers is not in the document]
Comment (darkwing) [reply src/reducers/tabs.js]: "Misusing"?  What are we misusing?  Suggested change?  Thank you for clarification!

[the remark this answers is not in the document]
Comment (AnshulMalik) [reply src/reducers/tabs.js]: Just the name `update` is misleading for both adding and updating.

**Pull request merged**